### ISSUE TITLE
Train

### ISSUE THREAD

**Issue by geraldstanje** [white-pony/go-fann#9]
Hello,

i dont want to read the training set from the file! i tried to manually train it using ann.Train, but the MSE does not go below 0.25. any idea why?
const numLayers = 3
const desiredError = 0.00001
const maxEpochs = 500000
const epochsBetweenReports = 1000

ann := fann.CreateStandard(numLayers, []uint32{2, 3, 1})
ann.SetActivationFunctionHidden(fann.SIGMOID_SYMMETRIC)
ann.SetActivationFunctionOutput(fann.SIGMOID_SYMMETRIC)

 trainInput := [][]fann.FannType{{-1.0, -1.0}, {-1.0, 1.0}, {1.0, -1.0}, {1.0, 1.0}}
 trainOutput := [][]fann.FannType{{-1.0}, {1.0}, {1.0}, {-1.0}}

  for i := 1; i <= maxEpochs; i++ {
    ann.ResetMSE()

```
for i, _ := range trainInput {
  ann.Train(trainInput[i], trainOutput[i])
}

if ann.GetMSE() < desiredError {
  break
}
```

  }

how can you fix that? could you please provide an example (maybe using TrainOnData)?

Thanks

**Comment by JohnNewbold** [comment]
Gas fees probably

**Comment by shanegammell47-dot** [comment]
i dont want to read the training set from the file! i tried to manually train it using ann.Train, but the MSE does not go below 0.25. any idea why?
const numLayers = 3
const desiredError = 0.00001
const maxEpochs = 500000
const epochsBetweenReports = 1000

ann := fann.CreateStandard(numLayers, []uint32{2, 3, 1})
ann.SetActivationFunctionHidden(fann.SIGMOID_SYMMETRIC)
ann.SetActivationFunctionOutput(fann.SIGMOID_SYMMETRIC)

trainInput := [][]fann.FannType{{-1.0, -1.0}, {-1.0, 1.0}, {1.0, -1.0}, {1.0, 1.0}}
trainOutput := [][]fann.FannType{{-1.0}, {1.0}, {1.0}, {-1.0}}

for i := 1; i <= maxEpochs; i++ {
ann.ResetMSE()

for i, _ := range trainInput {
  ann.Train(trainInput[i], trainOutput[i])
}

if ann.GetMSE() < desiredError {
  break
}
}

how can you fix that? could you please provide an example (maybe using TrainOnData)?

Thanks

The MSE is stuck around 0.25 because you are doing pure online training with a very small dataset and default FANN settings, so the gradient steps keep “bouncing” around a good solution instead of converging tightly; using `TrainOnData` (batch-style training) with a `TrainData` object and letting FANN manage epochs typically fixes this for XOR. [1][2]

### What is going wrong in your loop

In your code you are:

- Calling `ann.ResetMSE()` at the start of every epoch. [1]
- Training one sample at a time with `ann.Train(...)` (online training). [1]
- Checking `ann.GetMSE()` after each pass over 4 samples. [1]

For very small problems like XOR, online training plus default learning parameters often converges only to a modest error plateau, especially with symmetric sigmoids in both layers. FANN’s own examples use `TrainOnData` (or `TrainOnFile`) with many epochs and a relatively loose desired error (e.g. 0.001) and reach a good solution reliably. [3][2]

### How to fix it with TrainOnData

You can wrap your in‑memory data into a `TrainData` object and then use `TrainOnData`, which lets FANN handle epochs and MSE internally. [1][2]

In Go (using the fann-go style API), it looks like this:

```go
package main

import (
    "github.com/white-pony/go-fann"
)

func main() {
    numLayers := uint(3)
    desiredError := fann.FannType(0.0001)
    maxEpochs := uint(500000)
    epochsBetweenReports := uint(1000)

    ann := fann.CreateStandard(numLayers, []uint32{2, 3, 1})
    defer ann.Destroy()

    ann.SetActivationFunctionHidden(fann.SIGMOID_SYMMETRIC)
    ann.SetActivationFunctionOutput(fann.SIGMOID_SYMMETRIC)

    // XOR data in [-1,1] form
    trainInput := [][]fann.FannType{
        {-1, -1},
        {-1, 1},
        {1, -1},
        {1, 1},
    }
    trainOutput := [][]fann.FannType{
        {-1},
        {1},
        {1},
        {-1},
    }

    // Create training data object
    trainData := fann.CreateTrain(len(trainInput), 2, 1)
    defer trainData.Destroy()

    for i := 0; i < len(trainInput); i++ {
        trainData.SetInput(i, trainInput[i])
        trainData.SetOutput(i, trainOutput[i])
    }

    // Optional: initialize weights based on training data
    ann.InitWeights(trainData)

    // Train until desiredError or maxEpochs
    ann.TrainOnData(trainData, maxEpochs, epochsBetweenReports, desiredError)

    // Test network on training data
    for i := range trainInput {
        out := ann.Run(trainInput[i])
        println(trainInput[i][0], trainInput[i][1], "->", out[0])
    }
}
```

This mirrors the standard XOR examples that use `train_on_data` / `train_on_file` and successfully reach low MSE with SIGMOID_SYMMETRIC on both hidden and output layers. [3][2]

### If you really want manual training

If you insist on keeping the manual loop, you can help convergence by:

- Calling `ann.TrainEpoch(trainData)` instead of looping `ann.Train(...)` yourself, then reading `GetMSE()`. [1]
- Or, at least periodically evaluating error with `TestData` / `Test` on all 4 samples rather than relying only on the online‑updated MSE. [1]

But the simplest reliable fix for your case is to switch to `TrainOnData` as in the example above.

Citations:
[1] FANN Training https://libfann.github.io/fann/docs/files/fann_train-h.html
[2] AI::FANN(3pm) — libai-fann-perl - bookworm https://manpages.debian.org/bookworm/libai-fann-perl/AI::FANN.3pm.en.html
[3] Getting started with Neural Networks using the FANN library, PHP ... https://geekgirljoy.wordpress.com/2016/07/24/getting-started-with-neural-networks-using-the-fann-library-php-and-c9-io-part-2/
[4] AI::FANN - Perl wrapper for the Fast Artificial Neural Network library https://metacpan.org/pod/AI::FANN
[5] How continue training using library Fann? · Issue #102 - GitHub https://github.com/libfann/fann/issues/102
[6] Ubuntu Manpage: AI::FANN https://manpages.ubuntu.com/manpages/jammy/man3/AI::FANN.3pm.html
[7] Why did the error of the result of my ANN program only decreased a little while i was training it? https://stackoverflow.com/questions/65757009/why-did-the-error-of-the-result-of-my-ann-program-only-decreased-a-little-while
[8] Code for retraining of ANN gives always the same results (MSE) https://kr.mathworks.com/matlabcentral/answers/16697-code-for-retraining-of-ann-gives-always-the-same-results-mse
[9] FANN XOR training - c++ - Stack Overflow https://stackoverflow.com/questions/20776191/fann-xor-training
[10] Should MSE be able to get to 0 on training data with no noise? https://www.reddit.com/r/learnmachinelearning/comments/tpo3xv/should_mse_be_able_to_get_to_0_on_training_data/

**Comment by jasser2135-spec** [comment]
7402992575

**Comment by jasser2135-spec** [comment]
2637438616
